### ISSUE TITLE
Haskell-ghc-mod: ghc-mod failed to launch. It is probably missing or misconfigured. ENOENT

### ISSUE THREAD

**Issue by YetAnotherScrub** [atom-haskell/haskell-ghc-mod#179]
Apologies if I am just wasting time. New to Atom and github. Just installed Atom with a few Haskell packages and got the following error. Maybe there is an easy fix but I figure something is wrong if I get this on a fresh install. Relevant info should be in the report I guess.

[Enter steps to reproduce below:]

1. New Atom install today (downloaded installer from web).
2. Installed Haskell language, ghc-mod and Haskell ide (from the packages menu in Atom). ghc-mod didn't seem to work, and on restart got this error.

**Atom Version**: 1.10.2
**Electron Version**: 0.37.8
**System**: Microsoft Windows 10 Home
**Thrown From**: [haskell-ghc-mod](https://github.com/atom-haskell/haskell-ghc-mod) package, v1.17.2

### Stack Trace

Haskell-ghc-mod: ghc-mod failed to launch. It is probably missing or misconfigured. ENOENT

```
At Error: spawn ghc-mod ENOENT
PATH: C:\Program Files (x86)\Python\Scripts\;C:\Program Files (x86)\Python\;C:\Program Files (x86)\NVIDIA Corporation\PhysX\Common;C:\ProgramData\Oracle\Java\javapath;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Program Files (x86)\Skype\Phone\
path: C:\Program Files (x86)\Python\Scripts\;C:\Program Files (x86)\Python\;C:\Program Files (x86)\NVIDIA Corporation\PhysX\Common;C:\ProgramData\Oracle\Java\javapath;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Program Files (x86)\Skype\Phone\
Path: C:\Program Files (x86)\Python\Scripts\;C:\Program Files (x86)\Python\;C:\Program Files (x86)\NVIDIA Corporation\PhysX\Common;C:\ProgramData\Oracle\Java\javapath;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Program Files (x86)\Skype\Phone\

Error
    at file:///C:/Users/antho/.atom/packages/haskell-ghc-mod/lib/util.coffee:60:26
    at exithandler (child_process.js:221:5)
    at ChildProcess.errorhandler (child_process.js:233:5)
    at emitOne (events.js:90:13)
    at ChildProcess.emit (events.js:182:7)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:202:12)
    at onErrorNT (internal/child_process.js:348:16)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9)
```

### Commands

```
```

### Config

```json
{
  "core": {},
  "haskell-ghc-mod": {
    "highlightTooltips": false
  }
}
```

### Installed Packages

```coffee
# User
autocomplete-haskell, v0.6.7 (active)
haskell-ghc-mod, v1.17.2 (active)
ide-haskell, v1.8.1 (active)
language-haskell, v1.9.9 (active)
atom-dark-syntax, v0.27.0 (inactive)
atom-dark-ui, v0.52.0 (inactive)
atom-light-syntax, v0.28.0 (inactive)
atom-light-ui, v0.44.0 (inactive)
base16-tomorrow-dark-theme, v1.1.0 (inactive)
base16-tomorrow-light-theme, v1.1.1 (inactive)
one-dark-ui, v1.5.0 (active)
one-light-ui, v1.5.0 (inactive)
one-dark-syntax, v1.3.0 (active)
one-light-syntax, v1.3.0 (inactive)
solarized-dark-syntax, v1.0.2 (inactive)
solarized-light-syntax, v1.0.2 (inactive)
about, v1.5.3 (active)
archive-view, v0.61.1 (active)
autocomplete-atom-api, v0.10.0 (active)
autocomplete-css, v0.11.2 (active)
autocomplete-html, v0.7.2 (active)
autocomplete-plus, v2.31.1 (active)
autocomplete-snippets, v1.11.0 (active)
autoflow, v0.27.0 (inactive)
autosave, v0.23.1 (active)
background-tips, v0.26.1 (active)
bookmarks, v0.41.0 (active)
bracket-matcher, v0.82.1 (active)
command-palette, v0.38.0 (inactive)
deprecation-cop, v0.54.1 (active)
dev-live-reload, v0.47.0 (active)
encoding-selector, v0.22.0 (active)
exception-reporting, v0.39.0 (active)
find-and-replace, v0.201.0 (inactive)
fuzzy-finder, v1.3.0 (active)
git-diff, v1.1.0 (active)
go-to-line, v0.31.0 (inactive)
grammar-selector, v0.48.1 (active)
image-view, v0.58.2 (active)
incompatible-packages, v0.26.1 (active)
keybinding-resolver, v0.35.0 (active)
line-ending-selector, v0.5.0 (active)
link, v0.31.1 (inactive)
markdown-preview, v0.158.0 (active)
metrics, v0.53.1 (active)
notifications, v0.65.0 (active)
open-on-github, v1.2.0 (inactive)
package-generator, v1.0.0 (inactive)
settings-view, v0.242.2 (active)
snippets, v1.0.2 (active)
spell-check, v0.67.1 (active)
status-bar, v1.4.1 (active)
styleguide, v0.47.0 (active)
symbols-view, v0.113.0 (inactive)
tabs, v0.100.2 (active)
timecop, v0.33.2 (active)
tree-view, v0.208.2 (active)
update-package-dependencies, v0.10.0 (active)
welcome, v0.34.0 (active)
whitespace, v0.33.0 (active)
wrap-guide, v0.38.1 (active)
language-c, v0.52.1 (active)
language-clojure, v0.21.0 (active)
language-coffee-script, v0.47.2 (active)
language-csharp, v0.12.1 (active)
language-css, v0.37.1 (active)
language-gfm, v0.88.0 (active)
language-git, v0.15.0 (active)
language-go, v0.42.1 (active)
language-html, v0.45.1 (active)
language-hyperlink, v0.16.0 (active)
language-java, v0.23.0 (active)
language-javascript, v0.119.0 (active)
language-json, v0.18.2 (active)
language-less, v0.29.5 (active)
language-make, v0.22.2 (active)
language-mustache, v0.13.0 (active)
language-objective-c, v0.15.1 (active)
language-perl, v0.35.0 (active)
language-php, v0.37.2 (active)
language-property-list, v0.8.0 (active)
language-python, v0.45.0 (active)
language-ruby, v0.69.0 (active)
language-ruby-on-rails, v0.25.0 (active)
language-sass, v0.56.0 (active)
language-shellscript, v0.22.4 (active)
language-source, v0.9.0 (active)
language-sql, v0.23.0 (active)
language-text, v0.7.1 (active)
language-todo, v0.28.0 (active)
language-toml, v0.18.0 (active)
language-xml, v0.34.9 (active)
language-yaml, v0.26.0 (active)

# Dev
No dev packages
```

**Comment by lierdakil** [comment]
Please take a look at https://github.com/atom-haskell/ide-haskell#requirements (and possibly https://github.com/atom-haskell/haskell-ghc-mod#configuration) -- from what I can gather by looking at the debug log, you either didn't install ghc-mod binary, or didn't configure `haskell-ghc-mod` package. Also if you plan on using `stack`, take a look at https://github.com/atom-haskell/haskell-ghc-mod/wiki/Using-with-stack.

**Comment by YetAnotherScrub** [comment]
I set the Ghc mod path to C:\Users\[me]\.atom\packages\haskell-ghc-mod. That was the only setting that the configuration link says I should need to do. Given that the stack trace references it, I guess Atom has found the package itself fine.

**Comment by lierdakil** [comment]
I'm sorry, but did you actually read references provided? `haskell-ghc-mod` ***package*** relies on [`ghc-mod`](https://github.com/DanielG/ghc-mod) ***executable***, which you need to build. Which I assume you didn't do. This is explained in detail in the first link provided in my previous comment.

**Comment by YetAnotherScrub** [comment]
Ah I see now. Reading and understanding are 2 very different things. My apologies for expecting that a package would come with its core dependencies. I think I'm simply going to leave this for now, since I don't actually need it. Sorry to waste your time.

**Comment by lierdakil** [comment]
FYI, if core dependencies were integrate-able into the package, there would be no problem with that. But ghc-mod requires a whole other compiler stack (namely, GHC and cabal at the very least). Bundling that with a relatively tiny javascript editor package is borderline insanity (and besides, very error-prone). Haskell ecosystem is at its infancy, so some tradeoffs have to be made. As much as we'd like to provide a seamless experience with little to no hassle, it's just not possible at the moment.

Sorry if tone in my comments is a little harsh, I mean no offence. Just had a very rough month with very little spare time, and some of that fatigue might seep into my comments.

**Comment by YetAnotherScrub** [comment]
That's fine. I'm clearly clueless about this. Come to think of it, I should have realised that "ghc-mod" isn't just a small script to make the IDE package work.

**Comment by veehz** [comment]
@lierdakil Hi, I just started using Haskell in Atom and I've downloaded all required things, including the ghc-mod. But, Atom showed me this... How can I solve this?

[Enter steps to reproduce below:]

1. ...
2. ...

**Atom Version**: 1.10.2
**Electron Version**: 0.37.8
**System**: Ubuntu 16.04.1
**Thrown From**: [haskell-ghc-mod](https://github.com/atom-haskell/haskell-ghc-mod) package, v1.17.2


### Stack Trace

Haskell-ghc-mod: ghc-mod failed to launch. It is probably missing or misconfigured. ENOENT

```
At Error: spawn ghc-mod ENOENT
PATH: /home/mathnasium/bin:/home/mathnasium/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
path: undefined
Path: undefined

Error
    at /home/mathnasium/.atom/packages/haskell-ghc-mod/lib/util.coffee:60:26
    at exithandler (child_process.js:221:5)
    at ChildProcess.errorhandler (child_process.js:233:5)
    at emitOne (events.js:90:13)
    at ChildProcess.emit (events.js:182:7)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:202:12)
    at onErrorNT (internal/child_process.js:348:16)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9)
```

### Commands

```
```

### Config

```json
{
  "core": {},
  "haskell-ghc-mod": {}
}
```

### Installed Packages

```coffee
# User
autocomplete-haskell, v0.6.7 (active)
haskell-ghc-mod, v1.17.2 (active)
ide-haskell, v1.8.1 (active)
ide-haskell-cabal, v1.6.4 (active)
language-haskell, v1.9.9 (active)
atom-dark-syntax, v0.27.0 (inactive)
atom-dark-ui, v0.52.0 (inactive)
atom-light-syntax, v0.28.0 (inactive)
atom-light-ui, v0.44.0 (inactive)
base16-tomorrow-dark-theme, v1.1.0 (inactive)
base16-tomorrow-light-theme, v1.1.1 (inactive)
one-dark-ui, v1.5.0 (active)
one-light-ui, v1.5.0 (inactive)
one-dark-syntax, v1.3.0 (active)
one-light-syntax, v1.3.0 (inactive)
solarized-dark-syntax, v1.0.2 (inactive)
solarized-light-syntax, v1.0.2 (inactive)
about, v1.5.3 (active)
archive-view, v0.61.1 (active)
autocomplete-atom-api, v0.10.0 (active)
autocomplete-css, v0.11.2 (active)
autocomplete-html, v0.7.2 (active)
autocomplete-plus, v2.31.1 (active)
autocomplete-snippets, v1.11.0 (active)
autoflow, v0.27.0 (inactive)
autosave, v0.23.1 (active)
background-tips, v0.26.1 (active)
bookmarks, v0.41.0 (active)
bracket-matcher, v0.82.1 (active)
command-palette, v0.38.0 (inactive)
deprecation-cop, v0.54.1 (active)
dev-live-reload, v0.47.0 (active)
encoding-selector, v0.22.0 (active)
find-and-replace, v0.201.0 (inactive)
fuzzy-finder, v1.3.0 (active)
git-diff, v1.1.0 (active)
go-to-line, v0.31.0 (inactive)
grammar-selector, v0.48.1 (active)
image-view, v0.58.2 (active)
incompatible-packages, v0.26.1 (active)
keybinding-resolver, v0.35.0 (active)
line-ending-selector, v0.5.0 (active)
link, v0.31.1 (inactive)
markdown-preview, v0.158.0 (active)
notifications, v0.65.0 (active)
open-on-github, v1.2.0 (inactive)
package-generator, v1.0.0 (inactive)
settings-view, v0.242.2 (active)
snippets, v1.0.2 (active)
spell-check, v0.67.1 (active)
status-bar, v1.4.1 (active)
styleguide, v0.47.0 (active)
symbols-view, v0.113.0 (inactive)
tabs, v0.100.2 (active)
timecop, v0.33.2 (active)
tree-view, v0.208.2 (active)
update-package-dependencies, v0.10.0 (active)
welcome, v0.34.0 (active)
whitespace, v0.33.0 (active)
wrap-guide, v0.38.1 (active)
language-c, v0.52.1 (active)
language-clojure, v0.21.0 (active)
language-coffee-script, v0.47.2 (active)
language-csharp, v0.12.1 (active)
language-css, v0.37.1 (active)
language-gfm, v0.88.0 (active)
language-git, v0.15.0 (active)
language-go, v0.42.1 (active)
language-html, v0.45.1 (active)
language-hyperlink, v0.16.0 (active)
language-java, v0.23.0 (active)
language-javascript, v0.119.0 (active)
language-json, v0.18.2 (active)
language-less, v0.29.5 (active)
language-make, v0.22.2 (active)
language-mustache, v0.13.0 (active)
language-objective-c, v0.15.1 (active)
language-perl, v0.35.0 (active)
language-php, v0.37.2 (active)
language-property-list, v0.8.0 (active)
language-python, v0.45.0 (active)
language-ruby, v0.69.0 (active)
language-ruby-on-rails, v0.25.0 (active)
language-sass, v0.56.0 (active)
language-shellscript, v0.22.4 (active)
language-source, v0.9.0 (active)
language-sql, v0.23.0 (active)
language-text, v0.7.1 (active)
language-todo, v0.28.0 (active)
language-toml, v0.18.0 (active)
language-xml, v0.34.9 (active)
language-yaml, v0.26.0 (active)

# Dev
No dev packages
```

**Comment by lierdakil** [comment]
@veehz, not entirely sure what you mean by "downloaded". You have to build ghc-mod with your Haskell compiler, then either place resulting binaries on PATH (at the very least `ghc-mod` binary) or pointing Atom (via haskell-ghc-mod package settings) to the directory that contains said binaries (preferably through 'Additional Path Directories' setting). See ide-haskell's readme for more detailed instructions on building ghc-mod.

**Comment by veehz** [comment]
Thanks @lierdakil . My problem is solved now.

**Comment by nasagfunaro** [comment]
My comment is really to echo the comment from YetAnotherScrub.   I have tried a few times to go through the recommended links to see if this is installed or that is installed.  I still don't have it working for me. I will eventually get back and read the instructions again and see what I missed. I'll get it working. But I very much agree with YetAnotherScrub ,a package should be easier to install.  But what do I know, I have only about 45 years programming experience.

**Comment by lierdakil** [comment]
Look, I know it's frustrating, but.
complaining isn't really helping anyone.

I know  that installation seems a bit overcomplicated to new users, and
frankly it is. But as I outlined above, there's no easy fix for that at the
moment. Work is being done to alleviate that, but it's not something that
can be slapped together fast and pushed out into production, especially
considering that this project is short on hands.

In the meantime, I do my best to provide support. If that is not
satisfactory for you, you could look at alternatives, f. ex. Leksah or
emacs with Haskell-mode plugin or something else - there's quite a lot
actually. Bear in mind that there is no silver bullet, so you can run into
both similar and different challenges with getting alternatives working.

Or maybe you're feeling especially generous and could use your experience
to suggest solutions to the problem. That would be terribly nice of you,
seriously.

TL;DR:
If you need help - ask for it explicitly, and attach Atom's error report if
there is one.
If you want to help - that is always welcome.
If neither, and you're just here to vent some frustration - that's honestly
fine too, but kinda counterproductive.

**Comment by YetAnotherScrub** [comment]
Nasagfunaro if your problem is similar to mine it will be one of 3 things. (I'm writing this for anyone else who comes across this error, so I will assume very little knowledge.)
Essentially (as I now understand) the Atom package haskell-ghc-mod depends on having at least the basic Haskell compiler GHC installed on your system, and a thing called ghc-mod. If i understand correctly, there should be a big red flag here, because these are NOT an Atom things at all. You cannot get them through Atom.

Firstly you will need GHC. You probably have that of you are coding in Haskell, so I won't explain how to get it, but if anyone reading this didn't have that, search Google for "install Haskell ghc" and decide exactly what you want yourself.
Secondly you will need ghc-mod, which as the name suggests is a modification to the above. Although I don't think it changes the expected results of compiling Haskell code, you should consider whether you really want this. As a student whose work needs to be tested on other machines, I decided against it. If you do decide that you want it, then you install it from the terminal/command line using either cabal or stack, depending on what you installed with GHC. Further instructions are in the links which lierdakil kindly posted above. You may encounter issues, but resolving them is beyond the scope of this thread.
Thirdly, once you have ghc-mod, you will need to specify to the Atom package haskell-ghc-mod where it is installed. Do this from the package settings within Atom. It is recommended that you add the location to your PATH - if anybody doesn't know what that is or how to use it, Google is your friend - or you can tell Atom to use its own custom path and you can add the location of ghc-mod there.

Hopefully this will fix most issues of this kind. If anyone is certain that they have done all of these things and still gets the error, post an Atom report here. I agree that this is very complicated. However it is necessarily complicated, because ghc-mod is not a small change and certainly beyond the scope of what is basically a text editor to handle alone. My only complaint is that there should be a very obvious indicator when glancing at haskell-ghc-mod that it depends on external binaries, which the user must install on their own.

**Comment by lierdakil** [comment]
> Secondly you will need ghc-mod, which as the name suggests is a modification to the above. 

Actually, ghc-mod is a wrapper around GHC API that makes actually querying it for source information feasible, with the only reasonably stable alternative being ghci (which is not nearly as feature-rich, and is really awkward to work with from code). To quote ghc-mod's description:

> ghc-mod is a backend program to enrich Haskell programming in editors. It strives to offer most of the features one has come to expect from modern IDEs in any editor.

So this point:

>  As a student whose work needs to be tested on other machines, I decided against it.

doesn't really make much sense.

That said, you don't *need* neither ghc-mod nor haskell-ghc-mod package to use ide-haskell with other atom-haskell packages (except autocomplete-haskell that is -- it mostly depends on haskell-ghc-mod). You'll just be missing out on interactive features like context-aware autocompletion, type querying, etc. That's the primary reason why there's a whole lot of separate packages instead of one.

> My only complaint is that there should be a very obvious indicator when glancing at haskell-ghc-mod that it depends on external binaries

Well, I'm not sure what I can do besides stating that in the package readme. Which I did. Several times actually...

**Comment by YetAnotherScrub** [comment]
Well thanks for the classification on what it is. I figure it probably wouldn't have been a problem, but I want certain so didn't want to risk it.
On the last point, I know it's probably beyond your power but what I envision would be a feature of Atom packages where they have a tag that can be seen even just when searching packages, that indicates that the package is either fully self-contained, depends on other Atom packages, or depends on external binaries to fully function.
Alternatives would be having the package installer tell the user that they need these binaries for it to work (maybe not possible), or having "THIS PACKAGE DEPENDS ON EXTERNAL BINARIES" at the start of the description. (Possible but stupid)
Overall I see you are in a no-win situation until further notice. I dare say many more people will come along with the same silly problem.

**Comment by nasagfunaro** [comment]
Nikolay

Thank you so much for reaching out.  You are obviously a conscientious programmer.  You obviously want your product to work. I did not realize you interpret my comment as complaining. It was not intended to be a complaint.

Let me offer some observations from my end. These are just observations and please do not take anything personally. I would prefer that you know that for me I think your product is worth pursuing. The Haskell environment is a new environment for me. I stated earlier I have been programming for 45 years, so I am not intimidated by a new programming environments or things that don’t work as easily as one might hope. The Haskell environment is not a high priority so it is only getting a little of my attention. By the way I work on a Ubuntu, Mac and windows. So I can test across OS.
Once I figure it out I will be happy to share what I discovered.

I wish you the best.  If I ever get it working, I will be happy to let you know what worked for me.


1)       I am not frustrated. Bad assumption. I wouldn’t make assumptions on a user’s emotional state. ;-)  The problem does not seem to be with your code, but how ghc-mod is inconsistently loaded on different OS and different platforms. I am wading my way through Haskell–Platform, cabal and stack. I’ll get it figured out. I always do. Not a high priority for me.

2)       I am not a new user. Bad assumption. After 45 years of programming I have run across many programmers with comments similar to “you must be doing something wrong.” I try to never use those phrases. My philosophy is that users have the right to expect something to work smoothly, if it doesn’t work smoothly they lose faith in the code they downloaded. Regardless of how conscientious the provider (such as you) might be.  I am having difficulty getting Haskell IDE to work under emacs as well. I would hate to jump over to VIM. I realize none of this has anything to do with the code you provided.

3)       You are right, there does not appear to be an easy fix.  On my Mac OS 10.11, I completely removed the Haskell platform and tried to “home brew” a Haskell development environment this morning. “cabal install ghc-mod” failed for me.  That certainly is not your fault.  And I am wading thorugh the error messages when this email came through.

4)       Ah, the alternatives. I looked at them.  Good suggestions, and I have looked at them. My preference would be to get Atom and Haskell working together. I prefer the environment.  It does not appear to be your code.

So the next person that appears frustrated, I would say some to the effect: I realize it is not working on all platforms and I am currently working to resolve those issues by gaining insight from users like you. Offer some alternatives.

Sincerely and best regards
Greg Funaro

**Comment by lierdakil** [comment]
> It was not intended to be a complaint.

Sorry for misreading it then.

> I am not frustrated

I only intended to say that I understand if you are. Didn't mean to imply.

>  I am not a new user

Well, new user of ide-haskell -- that much seems to be true. Didn't mean to imply you're lacking experience with computers or anything else for that matter.

Also I should mention that after plowing your way through the mess of an ecosystem that Haskell is the first time around, subsequent dives become much easier -- primarily because you're already reasonably familiar with caveats and gotchas.

> “cabal install ghc-mod” failed for me

You might consider giving [stack](http://haskellstack.org) a try. While unnecessarily rigid in some cases, it can handle ghc installation for you, and is generally better than cabal at managing dependencies (mostly thanks to curated package repository) -- which makes dealing with what we call ‘Cabal hell’ much more bearable. There's a learning curve related to how different resolvers (i.e. package lists) relate to GHC versions, but it's relatively straightforward. One caveat is that you need to have one ghc-mod installation per resolver, and resolver-specific ghc-mod versions do lag behind, so you might not always get the up-to-date experience.

**Comment by nasagfunaro** [comment]
Thank you Nikolay.  You are correct the Haskell ecosystem is new to me. And your comments are helpful.

**Comment by cmkay56** [comment]
http://www.mew.org/~kazu/proj/ghc-mod/en/install.html

**Comment by crypticmind** [comment]
It wasn't working for me because I tried adding `$HOME/...` to the path. After actually reading the README, I changed it to `/Users/[me]/...` and it worked.

**Comment by bklebe** [comment]
echoing @crypticmind, the problem for me was using $HOME instead of the absolute path
